first_task.scala

```diff
@@ -0,0 +1,260 @@
+  .withColumn("creationDate", col("creationDate").cast(DateType))
+  .na.drop(Seq("creationDate"))
+
+val questionCols = Seq(
+  lit(null).alias("answerId"),
+  col("ownerUserId"),
+  col("creationDate").cast(DateType),
+  col("id").alias("questionId")
+)
+val answerCols = Seq(
+  col("id").alias("answerId"),
+  col("ownerUserId"),
+  col("creationDate").cast(DateType),
+  lit(null).alias("questionId")
+)
+val unionDf = questionNaFreeDs.select(questionCols: _*).unionAll(answersNaFreeDs.select(answerCols: _ *))
+val dailyStatisticsDf = unionDf
+  .groupBy("creationDate")
+  .agg(
+  countDistinct("ownerUserId").alias("unique_users"),
+  count("questionId").alias("number_of_questions"),
+  count("answerId").alias("number_of_answers")
+)
+
+dbutils.fs.rm("/FileStore/tables/daily.parquet", true) 
+dailyStatisticsDf.write.parquet("/FileStore/tables/daily.parquet")
+dailyStatisticsDf.show()
+
+// COMMAND ----------
+
+def countMeanAndMedianOfAnswerScores(dataset: Dataset[Answer], predicate: Answer => Boolean): (Double, Double) = {
+    val mean = dataset
+      .filter(predicate)
+      .select(avg("score"))
+      .take(1)(0)(0).toString().toDouble
+
+    val median = dataset.filter(predicate).stat.approxQuantile("score", Array(0.5), 0.1)(0)
+
+    (mean, median)
+}
+
+// COMMAND ----------
+
+// MAGIC %md
+// MAGIC What is the average and median score of the answers containing references (links) vs. those which don't have any refs(links)?
+
+// COMMAND ----------
+
+val containsLink = (a: Answer) => !Jsoup.parse(a.body).select("a[href]").isEmpty
+println(countMeanAndMedianOfAnswerScores(answersDs,containsLink(_)))
+println(countMeanAndMedianOfAnswerScores(answersDs, !containsLink(_)))
+
+// COMMAND ----------
+
+// MAGIC %md
+// MAGIC What is the average and median score of the answers containing code snippets against those who don't have any code snippets?
+
+// COMMAND ----------
+
+val containsLink = (a: Answer) => !Jsoup.parse(a.body).select("code").isEmpty
+println(countMeanAndMedianOfAnswerScores(answersDs,containsLink(_)))
+println(countMeanAndMedianOfAnswerScores(answersDs, !containsLink(_)))
+
+// COMMAND ----------
+
+// MAGIC %md
+// MAGIC Code length (lines) distribution (visualize)
+
+// COMMAND ----------
+
+// MAGIC %md
+// MAGIC Распределение по графику похоже на геометрическое, в котором p --- вероятность, того, что user закончит вводить код(включая "не начинание ввода"). Проверка гипотезы будет добавлена 
+
+// COMMAND ----------
+
+val codeLengthProvider: (String => Int) = Jsoup.parse(_).select("code").html.length
+val codeLenDf = answersDs.select(col("body"))
+  .unionAll(questionsDs.select(col("body")))
+  .map(r => codeLengthProvider(r(0).toString))
+
+display(codeLenDf)
+
+
+// COMMAND ----------
+
+// MAGIC %md
+// MAGIC Crrelation between the code length and answer score (visualize)
+
+// COMMAND ----------
+
+val codeLenScoreDf = answersDs
+    .map(a => (a.score, Jsoup.parse(a.body).select("code").html.length))
```

Comment:
```map()``` is not something used extensively in production spark for calculating values, replace with withColumn or similar. Also _1 and _2 is not nice.